src/concrete/ml/quantization/quantized_ops.py

```diff
@@ -554,22 +559,32 @@ def q_impl(
         assert q_input_1.quantizer.scale is not None
         assert q_input_1.quantizer.zero_point is not None
 
-        # De-quantize with input params and re-quantize with output parameters
-        # This will use TLUs over each element of the two inputs
-        # We do the de-quantization directly, instead of q_inputs[0].dequant(),
-        # So that we do not lose precision in the computation
+        # Dequantize
+        input_0 = q_input_0.dequant()
+        input_1 = q_input_1.dequant()
 
-        rescale_q0 = numpy.rint(
-            q_input_0.quantizer.scale
-            / self.output_quant_params.scale
-            * (q_input_0.qvalues + (-q_input_0.quantizer.zero_point))
-        ).astype(numpy.int64)
+        # If this operator is the last one in the graph,
+        # we rescale using the smallest scale to keep all information
+        if self.produces_graph_output:
+            common_scale = min(q_input_0.quantizer.scale, q_input_1.quantizer.scale)
+        # Otherwise we use the output op quantization scale
+        else:
+            common_scale = self.output_quant_params.scale
 
-        rescale_q1 = numpy.rint(
-            q_input_1.quantizer.scale
-            / self.output_quant_params.scale
-            * (q_input_1.qvalues + (-q_input_1.quantizer.zero_point))
-        ).astype(numpy.int64)
+        common_zero_point = 0
+        offset = 0
+
+        output_quant_params = UniformQuantizationParameters(
+            scale=common_scale,
+            zero_point=common_zero_point,
+            offset=offset,
+        )
+
+        quantizer = UniformQuantizer(params=output_quant_params, no_clipping=True)
```

Comment:
very nice adding `no_clipping`  and re-using the quantizer

Reply:
Got inspired by @RomanBredehoft code on linear models! Yes I think it's best we reuse all the tools we have created for quantizer so that we keep updating them to our needs.